LibWeb: Add OffscreenCanvas support

Totto16 opened the pull request:
This adds basic support for `OffscreenCanvas`.
It is just a skeleton with many complex methods stubbed out.
The more complicated things include:
- how to deal with the `skia_context` and everything related to that in WebWorkers
- adding (an optional) placeholder `HTMLCanvasElement (outside of the DOM), synchronize the state of that like it is specified in the spec
- support every operation with Window as global object and with Worker, this requires some existing functionality to be adapted to `OffscreenCanvas` as argument

There was one spec bug (https://github.com/whatwg/html/issues/11101) that the spec for `OffscreenCanvas` has.

Some of these things were already tackled, but not all of them. It works as a basic skeleton, everything else can be done in additional PRs

ADKaster commented on Libraries/LibWeb/HTML/OffscreenCanvas.cpp:
```diff
@@ -0,0 +1,380 @@
+    auto bitmap = bitmap_or_error.release_value();
+
+    // 1. Initialize the bitmap of this to a rectangular array of transparent black pixels of the dimensions specified by width and height.
+    // TODO: I am not sure if this is already implied by the initialization, since it is all 0s? Ask that in the GitHub PR
```
Question here

tcl3 replied:
The Bitmap's pixels are set to `rgba(0,0,0,0)` when the Bitmap is created, so I don't believe it's necessary to set them here.

ADKaster commented on Libraries/LibWeb/HTML/OffscreenCanvas.cpp:
```diff
@@ -0,0 +1,380 @@
+
+WebIDL::ExceptionOr<void> OffscreenCanvas::set_width(WebIDL::UnsignedLong value)
+{
+    if (value > 2147483647)
```
where does this check come from? is there a spec step for setting the width?

Otherwise, this should be expressed in terms of AK::NumericLimits. or a call to AK::clamp.

Totto16 replied:
It was a copy paste from the same code in HTMLCanvasElement 😓  The spec doesnt mention it as far as I know

ADKaster commented on Libraries/LibWeb/HTML/Canvas/SerializeBitmap.h:
```diff
@@ -0,0 +1,23 @@
+};
+
+// https://html.spec.whatwg.org/multipage/canvas.html#a-serialisation-of-the-bitmap-as-a-file
+ErrorOr<SerializeBitmapResult> serialize_bitmap(Gfx::Bitmap const& bitmap, StringView type, Optional<double> quality);
```
creating this separate file for serializing bitmaps should be done before deleting the code in the OffscreenCanvas commit. Ideally, as its own atomic commit.

ADKaster commented on Libraries/LibWeb/HTML/Canvas/SerializeBitmap.cpp:
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (c) 2025, Ladybird contributors
```
We don't have "Ladybird contributors" as a standard copyright line. It's encouraged to use your own name + email, or "the Ladybird developers." if you're not comfortable with that for some reason.

ADKaster commented on Libraries/LibWeb/HTML/Canvas/CanvasDrawImage.cpp:
```diff
@@ -42,6 +42,12 @@ static void default_source_size(CanvasImageSource const& image, float& source_wi
                 source_height = source->video_height();
             }
         },
+        [&source_width, &source_height](GC::Root<OffscreenCanvas> const& source) {
```
`LibWeb: Fix OffscreenCanvas Related errors` is very suspicious. If any of the previous commits don't build for any reason, the changes should be moved back. We want every commit to build and pass all tests in isolation.

ADKaster commented on Libraries/LibWeb/HTML/OffscreenCanvas.cpp:
```diff
@@ -292,14 +293,23 @@ GC::Ref<WebIDL::Promise> OffscreenCanvas::convert_to_blob(Optional<ImageEncodeOp
             } else {
                 // 1. If result is non-null, resolve result with a new Blob object, created in the relevant realm of this OffscreenCanvas object, representing file. [FILEAPI]
-                GC::Ptr<FileAPI::Blob> blob_result = FileAPI::Blob::create(realm(), file_result->buffer, MUST(vm(), String::from_utf8(file_result->mime_type)));
+                auto type = String::from_utf8(file_result->mime_type);
```
this change should be moved back to the commit that added the task_to_queue

ADKaster commented on Tests/LibWeb/Text/expected/all-window-properties.txt:
```diff
@@ -268,6 +268,8 @@ NodeList
 Number
 Object
 OfflineAudioContext
+OffscreenCanvas
```
this change too, move back.

ADKaster reviewed (commented):
I'd like either @Lubrsi  or @kalenikaliaksandr to take a look at this as well, but overall this seems like a good start. The commits need a bit of reshuffling, however.

Totto16 commented:
> I'd like either @Lubrsi or @kalenikaliaksandr to take a look at this as well, but overall this seems like a good start. The commits need a bit of reshuffling, however.

I have my own mental model on how atomic the commits should be, so what is the way you intend it to be. Everything in one is too big imo. The creation of IDL files, test in separate commits, the creation of a separate file for `SerializeBitmap` in one. But the rest?

ADKaster commented:
"how atomic" has a clear definition: the repository with commit as HEAD must build without warnings or errors. Every commit should pass all tests in the test suite.

However you chunk up the changes to achieve that is highly change-dependent.

ADKaster commented:
> So would one commit for all modification in the cpp files be too big in this case?

If you did that, the maintainers would start pointing out places where changes are independent, like your move of the buffer serialization code. Giant mono-commits are also harder to review, and might result in changes going unreviewed for a while until you poke #code-review enough.

tcl3 commented on Libraries/LibWeb/HTML/Canvas/CanvasTextDrawingStyles.h:
```diff
@@ -40,6 +42,34 @@ class CanvasTextDrawingStyles {
+    {
+        // Font resolution for the font style source object requires a font source. This is determined for a given object implementing CanvasTextDrawingStyles by the following steps: [CSSFONTLOAD]
+
+        if constexpr (std::is_same_v<CanvasType, HTML::HTMLCanvasElement>) {
```
`SameAs<>` should be used instead of `std::is_same_v<>` here

tcl3 commented on Libraries/LibWeb/HTML/Canvas/CanvasTextDrawingStyles.h:
```diff
@@ -59,12 +89,35 @@ class CanvasTextDrawingStyles {
+
+        auto font_list = font_source.visit(
+            [&](DOM::Document* document) -> RefPtr<Gfx::FontCascadeList const> {
+                if constexpr (std::is_same_v<CanvasType, HTML::HTMLCanvasElement>) {
```
`SameAs<>` should be used instead of `std::is_same_v<>` here

Totto16 commented:
> This pull request has been automatically marked as stale because it has not had recent activity. It will be closed in 7 days if no further activity occurs. Thank you for your contributions!

👀

Totto16 commented:
> Hi! Just popping in here to say that this PR seems to add ~200 new WPT passes in `html/canvas/offscreen`, but I also see an increase in crashes from 1 to 44. Definitely something to check out.

Yeah, the stubbed out things certainly cause some issues, maybe also crashes, but this PR is already large. 

I don't know, if you operate with feature branches, but maybe we could tackle that in that way. I am willing to work more on this. But doing so, without approval for this basic structure PR, is just a waste of time, since resolving later merge conflicts with future work is not that easy.

This PR is already quite large in its size, so I don't know what the best option is 🤷🏼‍♂️ 
I just know, that we have to start somewhere with the OffscreenCanvas, and if there are some crashes, that don't occur after a long time, I personally think it is okay, to merge it into main 🤷🏼‍♂️ 

But you (as in the maintainers) have to decide where to go with this, I am happy to help 😄

Totto16 commented:
Hi @gmta, I made #4726 to make the changes from #4506 (your PR) available to other files too, and not implemented inside `CanvasRenderingContext2D`. But I didn't fully test these changes and if they would be enough, but now I did that and found out, that `CanvasRenderingContext2D::context_attributes_from_options` (see [here](https://github.com/LadybirdBrowser/ladybird/pull/4506/files#diff-cfbd5212fabc9bc95bdf094f8763ecf02c039a9ef0cf8cf85e4d2d9594fa0d74R66-R111) ) also needs to be shared, I added a commit here, to move it into the `LibWeb/HTML/Canvas/CanvasSettings.*` files so that it can be used also from `OffscreenCanvasRenderingContext2D`. 

My question would be, should I make this commit (atm the moment of writing ea21480b5bab98e9406951f0b745a15a93cc4348, but might change hash, due to rebases) into a separate PR like #4726 or just keep it in this PR?

And also, I copied your copyright into the new file, as I only moved the content and changed the name, but the content is the same, is that fine for you?
